src/xmerl_dsig.erl

```diff
@@ -0,0 +1,730 @@
+    [NotElem | do_strip_signature(Tail, Name, Parents)];
+do_strip_signature([], _, _) ->
+    [].
+
+%% @doc Signs the given XML element by creating a ds:Signature element within it, returning
+%%      the element with the signature added.
+%%
+%% Don't use "ds" as a namespace prefix in the envelope document, or things will go baaaad.
+%% This function has now been modified to do Skatteverket's funky removing of
+%% all namespace prefixes before doing a digest. Also there is function
+%% that recreates bugs, which can be removed if you are not implementing
+%% the Skatteverket API.
+%% The private key is now hidden inside a fun so it isn't visible in crashes.
+%% It will now sign all signatures inside the SOAP structure.
+-spec sign(Element :: #xmlElement{},
+           PrivateKeyFun :: fun(() -> #'RSAPrivateKey'{}),
+           CertBin :: binary()) -> #xmlElement{}.
+sign(ElementIn, PrivateKeyFun, CertBin) when
+      is_binary(CertBin) ->
+    case xmerl_xpath:string(
+           "//*[\"Signature\"=local-name() and \"http://www.w3.org/2000/09/"
+           "xmldsig#\"=namespace-uri()]", ElementIn) of
+        [] ->
+            {error, no_signature};
+        Signatures when is_list(Signatures) ->
+            sign_signatures(lists:reverse(Signatures), ElementIn, PrivateKeyFun,
+                            CertBin)
+    end.
+
+sign_signatures([Signature|Tail], ElementIn, PrivateKeyFun, CertBin) ->
+    PrivateKey = #'RSAPrivateKey'{} = PrivateKeyFun(),
+    logger:debug("Signature: ~p~n", [Signature]),
+    TransformAlgo =
+        xmerl_xpath:string(
+          "//*[\"Transform\"=local-name() and \"http://www.w3.org/2000/09/"
+          "xmldsig#\"=namespace-uri()]/@Algorithm", Signature),
+    [#xmlAttribute{value = SignatureMethodAlgorithm}] =
+        xmerl_xpath:string(
+          "//*[\"SignatureMethod\"=local-name() and \"http://www.w3.org/2000/"
+          "09/xmldsig#\"=namespace-uri()]/@Algorithm", Signature),
+    {HashFunction, _DigestMethod, SignatureMethodAlgorithm} =
+        signature_props(SignatureMethodAlgorithm),
+
+    [#xmlAttribute{value = "http://www.w3.org/2001/10/xml-exc-c14n#"}] =
+        xmerl_xpath:string(
+          "//*[\"CanonicalizationMethod\"=local-name() and \"http://www.w3.org/"
+          "2000/09/xmldsig#\"=namespace-uri()]/@Algorithm", Signature),
+    [#xmlAttribute{value = SignatureMethodAlgorithm}] =
+        xmerl_xpath:string(
+          "//*[\"SignatureMethod\"=local-name() and \"http://www.w3.org/2000/"
+          "09/xmldsig#\"=namespace-uri()]/@Algorithm", Signature),
+    [_C14nTx = #xmlElement{}] =
+        xmerl_xpath:string(
+          "//*[\"Transform\"=local-name() and \"http://www.w3.org/2000/09/"
+          "xmldsig#\"=namespace-uri()]", Signature),
+         %% xmerl_xpath:string(
+         %%  "ds:Signature/ds:SignedInfo/ds:Reference/ds:Transforms/ds:Transform"
+         %%  "[@Algorithm='http://www.w3.org/2001/10/xml-exc-c14n#']",
+         %%  Signature, [{namespace, DsNs}]),
```

Comment:
What's the purpose of this unused match? Remove it or make an error with a cause that someone can interpret.